Expression parser for parameters

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
close #2348, #2603.


### Details and comments
AST based parser is implemented under `qiskit.tools`. This parser accepts limited operators (`+`, `-`, `*`, `/`, `**`) and limited functions (defined under `cmath`). And the parser returns a tuple of expression (callable) and list of parameter names. This parser has no dependency on python evaluation function.

```
expr = '-P1*cos(P2)'
parsed_expr, params = parse_string_expr(expr)

value = parsed_expr(P1=1.0, P2=2.0)
```

nkanazawa1989 commented on qiskit/tools/parser.py:
```diff
@@ -0,0 +1,149 @@
+_math_ops = {key: value for (key, value) in vars(cmath).items() if key[0] != '_'}
+
+# valid parameters
+_param_regex = r'(P\d+)'
```
I think I need to add some regular expression of parameters for circuit. #2375

taalexander commented on qiskit/tools/parser.py:
```diff
@@ -0,0 +1,149 @@
+
+
+# valid functions
+_math_ops = {key: value for (key, value) in vars(cmath).items() if key[0] != '_'}
```
I think this should be replaced with an explicit list of allowed functions from `cmath`.

taalexander commented on qiskit/tools/parser.py:
```diff
@@ -0,0 +1,149 @@
+    for match, sub in subs:
+        source = source.replace(match, sub)
+
+    params = sorted(re.findall(_param_regex, source))
```
Could the params now be parsed as part of the AST?

taalexander commented on qiskit/tools/parser.py:
```diff
@@ -0,0 +1,149 @@
+    Raises:
+        QiskitError: If expression is not parsable.
+    """
+    subs = [('numpy.', ''), ('np.', ''), ('math.', '')]
```
cmath should be appended to this list.

taalexander commented on qiskit/tools/parser.py:
```diff
@@ -0,0 +1,149 @@
+        elif isinstance(node, ast.Num):
+            # number
+            try:
+                return complex(node.n)
```
will this cause every evaluated expression to return a complex, even if the input number was a float?

nkanazawa1989 commented:
@taalexander thank you for reviewing! Creating a grammar for arithmetic expresions may introduce additional dependency, so I prefer ast. In e5f90c7fc33d3ac3387ca0177c80aaa048ae6ba0, I replaced recursive call of `_visit_node` by [ast.NodeTransformer](https://tobiaskohn.ch/index.php/2018/07/30/transformations-in-python/). The main reasons of this are:

> Could the params now be parsed as part of the AST? https://github.com/Qiskit/qiskit-terra/pull/2493#discussion_r287603348

Now params are parsed by AST as well. Arbitrary names for parameter are accepted. This is for parsing circuit parameters - such as theta, nu, etc...

> They should support partial binding, provide a reasonably display of their operations and parameters when included in a printed circuit https://github.com/Qiskit/qiskit-terra/issues/2375#issue-442228477

`QiskitExpression` inheriting from NodeTransformer returns `float` or `complex` when all parameters are bound, otherwise it returns `ast`. NodeTransformer overwrite a node by `ast.Num` if parameters in its child nodes are bound.

`parse_string_expr` returns tuple of `QiskitExpression` and list of parameter names.

We can get Expression as follows
```
expr = '-P1*cos(P2)'
expression, params = parse_string_expr(expr)

value = expression(P1=1.0, P2=2.0)
```
Also we can specify partial biding
```
node = expression(P1=1.0, allow_partial_bind=True)
```
This returns partially evaluated abstract syntax tree.

taalexander commented:
After looking through the parameter expression I think it should still be possible to integrate with the `ParameterExpression`. We do not need to support the entire list of built-ins as is currently done (pulse doesn't actually use these right now, most expressions are of the form `2*np.pi*P1`. If you were to define an expression `ops` module or something like this, we could then use this as the set of supported operations and as the AST is parsed, the expression could be dynamically built.

nkanazawa1989 commented:
I moved parser under `qiskit.pulse` and changed its name from `QiskitExpression` to `PulseExpression`.

This parser receives an expression in str format (from backend), and I can see some differences in the format of expression from #2537 . @kdk is it difficult to define expressions in str format?

@taalexander for partial biding, `functools.partial` is returned instead of AST. but it is difficult to define params by args now. I propose to delete `*args` from arguments of parser and leave only `**kwargs` to remove ambiguity.

```
expr = 'P1+P2+P3'
expression, params = parse_string_expr(expr, partial_binding=True)
```

example:
```
expr = expression(P1=1.0, P2=2.0)
value = expr(P3=3.0)
```
it works (returns 6.0).

```
expr = expression(1.0, 2.0)
value = expr(P3=3.0)
```
it works (returns 6.0).

```
expr = expression(P1=1.0, P2=2.0)
value = expr(3.0)
```
it overwrites P1 and returns `functools.partial` (because P3 is not defined).

taalexander commented on qiskit/pulse/parser.py:
```diff
@@ -162,7 +165,10 @@ def visit_Expression(self, node):
             ast.Expression: Evaluated value.
         """
-        return self.visit(node.body)
+        tmp_node = copy.deepcopy(node)
```
Could you please help me understand why copying is necessary now?

nkanazawa1989 replied:
Because it overwrites original ast node, we need to copy it before calling `visit`. `tmp_node` is returned and passed to create new parser object if parameters are partially bound. Otherwise, we cannot reuse a parser for a different parameter set.

nkanazawa1989 replied:
I wrote this test to check if the parser can be reused for another parameter.
https://github.com/Qiskit/qiskit-terra/blob/b0d9b0b247987b9b6356a07768ef7d4ee5e1e0cc/test/python/pulse/test_parser.py#L207-L215

taalexander commented on qiskit/pulse/parser.py:
```diff
@@ -17,141 +17,144 @@
+from qiskit.pulse.exceptions import PulseError
+
+
+class PulseExpression(ast.NodeTransformer):
```
It seems now that your view is that this will be pulse only? Do you see this becoming a general qiskit expression in the future?

nkanazawa1989 replied:
I think this can be a general parser for string input, but circuit parser receives more pythonic object as @kdk pointed, and I don't think to integrate circuit parser with this. If any string expression exists in circuit, this parser should be used.

taalexander replied:
Ok we can always move this in the future.

taalexander commented on qiskit/pulse/parser.py:
```diff
@@ -0,0 +1,292 @@
+
+    expression = PulseExpression(source, partial_binding)
+
+    return expression, expression.params
```
If `params` are already an attribute of `expression` maybe we should only just return `expression`?

nkanazawa1989 replied:
I agree. This is just to align with old interface. I'll update converter module to use params property.
